pydantic/json_schema.py

```diff
@@ -0,0 +1,1081 @@
+        return {'anyOf': members}
+
+    def get_json_ref_counts(self, json_schema: JsonSchemaValue) -> dict[JsonRef, int]:
+        """
+        Get all values corresponding to the key '$ref' anywhere in the json_schema
+        """
+        json_refs: dict[JsonRef, int] = Counter()
+
+        def _add_json_refs(schema: Any) -> None:
+            if isinstance(schema, dict):
+                if '$ref' in schema:
+                    json_ref = JsonRef(schema['$ref'])
+                    already_visited = json_ref in json_refs
+                    json_refs[json_ref] += 1
+                    if already_visited:
+                        return  # prevent recursion on a definition that was already visited
+                    _add_json_refs(self.definitions[self.json_to_defs_refs[json_ref]])
+                for v in schema.values():
+                    _add_json_refs(v)
+            elif isinstance(schema, list):
+                for v in schema:
+                    _add_json_refs(v)
+
+        _add_json_refs(json_schema)
+        return json_refs
+
+    def handle_invalid_for_json_schema(self, schema: CoreSchema | TypedDictField, error_info: str) -> JsonSchemaValue:
+        if CoreMetadataHandler(schema).get_modify_js_function():
+            # Since there is a json schema modify function, assume that this type is meant to be handled,
+            # and the modify function will set all properties as appropriate
+            return {}
+        else:
+            raise PydanticInvalidForJsonSchema(f'Cannot generate a JsonSchema for {error_info}')
+
+
+def schema(
+    models: Sequence[type[BaseModel] | type[Dataclass]],
+    *,
+    by_alias: bool = True,
+    title: str | None = None,
+    description: str | None = None,
+    ref_template: str = DEFAULT_REF_TEMPLATE,
+    schema_generator: type[GenerateJsonSchema] = GenerateJsonSchema,
+) -> dict[str, Any]:
+    instance = schema_generator(by_alias=by_alias, ref_template=ref_template)
+    definitions = instance.generate_definitions([get_model(x).__pydantic_core_schema__ for x in models])
+
+    json_schema: dict[str, Any] = {}
+    if definitions:
+        json_schema['$defs'] = definitions
+    if title:
+        json_schema['title'] = title
+    if description:
+        json_schema['description'] = description
+
+    return json_schema
+
+
+def model_schema(
+    model: type[BaseModel] | type[Dataclass],
+    by_alias: bool = True,
+    ref_template: str = DEFAULT_REF_TEMPLATE,
+    schema_generator: type[GenerateJsonSchema] = GenerateJsonSchema,
+) -> dict[str, Any]:
+    model = get_model(model)
+    return model.model_json_schema(by_alias=by_alias, ref_template=ref_template, schema_generator=schema_generator)
+
+
+# TODO: It might be a good idea to move these functions to a separate file, whether in pydantic_core or not
+#   Alternatively, might be good to make them "public"
+def _is_typed_dict_field(schema: CoreSchema | TypedDictField) -> TypeGuard[TypedDictField]:
```

Comment:
can we move these into `_internal`?

Reply:
Yeah, do you prefer them in pydantic instead of pydantic_core? I'll put it in an `_internal` module if in pydantic, just making sure.